---
title: Introduction to Single-cell RNA-seq Schedule
sidebar: false
toc: false
execute: 
  eval: false
---

# Introduction to single-cell RNA-seq

## Pre-reading

* [Introduction to scRNA-seq](../lessons/01_intro_to_scRNA-seq.qmd)
* [Raw data to count matrix](../lessons/02_SC_generation_of_count_matrix.qmd)
* [Download this project](https://www.dropbox.com/s/vop78wq76h02a2f/single_cell_rnaseq.zip?dl=1)

## Day 1

| Time |  Topic  | Instructor |
|:-----------:|:----------:|:--------:|
| 09:30 - 09:45 | [Workshop introduction](../slides/workshop_intro_slides.pdf) | Will |
| 09:45 - 10:35| [Introduction to Single Cell RNA-sequencing: a practical guide](../slides/072624_Chan_scRNAseq_workshop_Arpita_Kulkarni.pdf) | [Dr. Arpita Kulkarni](https://singlecellcore.hms.harvard.edu/people/arpita-kulkarni-phd) |
| 10:35 - 10:40 | Break |
| 10:40 - 11:00 | scRNA-seq pre-reading discussion | All |
| 11:00 - 11:45 | [Quality control set-up](../lessons/03_SC_quality_control-setup.ipynb) | Noor |
| 11:45 - 12:00 | Overview of self-learning materials and homework submission | Will |

### Before the next class:

I. Please **study the contents** and **work through all the code** within the following lessons:
 
   1. [Quality control of cellranger counts](../lessons/04_cellranger_QC.ipynb)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
       <br>Before you start any analysis, itâ€™s important to know whether or not you have good quality cells. At these early stages you can flag or remove samples that could produce erroneous results downstream. <br><br>In this lesson you will:<br>
             - Discuss the outputs of cellranger and how to run it <br>
             - Review web summary HTML report<br>
             - Create plots from metrics_summary.csv file <br><br>
        </details>

 2.  [Quality control with additional metrics](../lessons/04_SC_quality_control.ipynb)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
         <br>In addition to the QC generated by cellranger, we can also compute some of our own metrics based on the raw data we have loaded into our Seurat object. <br><br>In this lesson you will:<br>
             - Compute essential QC metrics for each sample<br>
             - Create plots to visualize metrics across cells per sample<br>
             - Critically evaluate each plot and learn what each QC metric means<br><br>
        </details>
     
   3. [Theory of PCA](../lessons/05_theory_of_PCA.qmd)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
         <br>Before we can begin the next steps of the workflow, we need to make sure you have a good understanding of Principal Components Analysis (PCA). This method will be utilized in the scRNA-seq analysis workflow, and this foundation will help you better navigate those steps and interpretation of results.<br><br>
        </details>
         

II. **Submit your work**:
   * Each lesson above contains exercises; please go through each of them.
   * **Submit your answers** to the exercises using [this Google form](https://docs.google.com/forms/d/e/1FAIpQLSf4zw37eUcuNCmHxCLikg3JENoxGWIuv5c_5pqHdYNId81Arg/viewform) on **the day *before* the next class**.
   


### Questions?
* ***If you get stuck due to an error*** while runnning code in the lesson, [email us](mailto:hbctraining@hsph.harvard.edu) 

***